### ISSUE TITLE
UpdateRemotePlugins fails with an exception

### ISSUE THREAD

**Issue by PonderingGrower** [neovim/neovim#3749]
While I try to use the [deoplete](https://github.com/Shougo/deoplete.nvim) plugin and use `UpdateRemotePlugins` I get the following exception.

    Vim(let):E903: Could not spawn API job
    Error detected while processing function <SNR>43_UpdateRemotePlugins..<SNR>43_RegistrationCommands..remote#host#Require..<SNR>43_RequirePythonHost:
    line   21:
    E605: Exception not caught: Failed to load Python host. You can try to see what happened by starting Neovim with the environment variable $NVIM_PYTHON_LOG_FILE set to a file and opening the generated log file. Also, the host stderr will be
     available in Neovim log, so it may contain useful information. See also ~/.nvimlog.
    Error detected while processing function <SNR>43_UpdateRemotePlugins..<SNR>43_RegistrationCommands..remote#host#Require:
    line   10:
    E171: Missing :endif
    Error detected while processing function <SNR>43_UpdateRemotePlugins:
    line    5:
    E171: Missing :endif

I have `$NVIM_PYTHON_LOG_FILE` set:

    :!env | grep NVIM
    NVIMRUNTIME=/usr/local/share/nvim/runtime
    NVIM_PYTHON_LOG_FILE=/tmp/nvim_python.log
    NVIM_LISTEN_ADDRESS=/tmp/nvim8zAmOq/0

But the file does not exist:

    zsh: no matches found: /tmp/nvim_python.log*

I have `python3` `neovim` package installed:

     $ sudo pip3 show neovim 
    ---
    Metadata-Version: 1.1
    Name: neovim
    Version: 0.0.38
    Summary: Python client to neovim
    Home-page: http://github.com/neovim/python-client
    Author: Thiago de Arruda
    Author-email: tpadilha84@gmail.com
    License: Apache
    Location: /usr/local/lib/python3.2/dist-packages
    Requires: msgpack-python, trollius, greenlet

What gives?

**Comment by PonderingGrower** [comment]
With the following minimal `init.nvim`:

    call plug#begin('~/.config/nvim/bundle')
    Plug 'Shougo/deoplete.nvim'
    call plug#end()
    let g:deoplete#enable_at_startup = 1

After running `UpdateRemotePlugins` I get:

    Vim(if):Channel was closed by the client
    Error detected while processing function <SNR>18_UpdateRemotePlugins..<SNR>18_RegistrationCommands..remote#host#Require..<SNR>18_RequirePythonHost:
    line   21:
    E605: Exception not caught: Failed to load Python host. You can try to see what happened by starting Neovim with the environment variable $NVIM_PYTHON_LOG_FILE set to a file and opening the generated log file. Also, the host stderr will be
     available in Neovim log, so it may contain useful information. See also ~/.nvimlog.
    Error detected while processing function <SNR>18_UpdateRemotePlugins..<SNR>18_RegistrationCommands..remote#host#Require:
    line   10:
    E171: Missing :endif
    Error detected while processing function <SNR>18_UpdateRemotePlugins:
    line    5:
    E171: Missing :endif

**Comment by PonderingGrower** [comment]
Oh, I just realized this might be an issue only related to [vim-plug](https://github.com/junegunn/vim-plug).

**Comment by PonderingGrower** [comment]
Ok, I made https://github.com/junegunn/vim-plug/issues/332. I will close this issue as I think this is only related to [vim-plug](https://github.com/junegunn/vim-plug). Sorry for the noise.

**Comment by PonderingGrower** [comment]
So I was informed on https://github.com/junegunn/vim-plug/issues/332 that `UpdateRemotePlugins` is indeed a `neovim` command and not from `vim-plug`. My mistake. I assume it is based on it not being available when using `nvim -u NONE`.

I will repeat what I just did to verify this issue.

 1. Removed `neovim` pip package,
 2. Installed it again.
 3. Created new empty `~/.config/nvim`
 4. Created this minimal `init.vim`

`init.vim`:

    call plug#begin('~/.config/nvim/bundle')
    "Plug 'Shougo/deoplete.nvim'
    Plug 'Shougo/unite.vim'
    call plug#end()

 5. Installed `unite` using `PlugInstall`.
 6. Resterted `neovim`
 7. Tried using `UpdateRemotePlugins`:

And I get:

    Vim(if):Channel was closed by the client
    Error detected while processing function <SNR>22_UpdateRemotePlugins..<SNR>22_RegistrationCommands..remote#host#Require..<SNR>22_RequirePythonHost:
    line   21:
    E605: Exception not caught: Failed to load Python host. You can try to see what happened by starting Neovim with the environment variable $NVIM_PYTHON_LOG_FILE set to a file and opening the generated log file. Also, the host stderr will be
     available in Neovim log, so it may contain useful information. See also ~/.nvimlog.
    Error detected while processing function <SNR>22_UpdateRemotePlugins..<SNR>22_RegistrationCommands..remote#host#Require:
    line   10:
    E171: Missing :endif
    Error detected while processing function <SNR>22_UpdateRemotePlugins:
    line    5:
    E171: Missing :endif

**Comment by jszakmeister** [comment]
That looks like Python support is missing.  Have you looked at [:help nvim-python](https://neovim.io/doc/user/nvim_python.html)?

**Comment by PonderingGrower** [comment]
Yes I have, and as I said I already have the `neovim` pip package installed:

Running `:echo has('python3')` gives me `1`.
Running `:py print 1+1` gives me `2`.

**Comment by PonderingGrower** [comment]
My package comes from this repo: https://launchpad.net/~neovim-ppa/+archive/ubuntu/unstable

    Package: neovim
    Status: install ok installed
    Priority: optional
    Section: editors
    Installed-Size: 19100
    Maintainer: Florian Walch <florian+launchpad@fwalch.com>
    Architecture: amd64
    Version: 0.1.0ubuntu1+git201511260052+2015+15~ubuntu12.04.1
    Provides: editor, vim
    Depends: libc6 (>= 2.14)
    Description: Fork of Vim aiming to improve user experience, plugins, and GUIs.
    Homepage: http://neovim.io

**Comment by PonderingGrower** [comment]
Oooooh, I see, [:help nvim-python](https://neovim.io/doc/user/nvim_python.html) says that Python has to be 3.3 and I have 3.2.

I think that's it. I'll try to find a way to get 3.3 on Ubuntu Precise and report back.

**Comment by jszakmeister** [comment]
Oh, that could be the issue.  Also, check if there is anything in `~/.nvimlog` (versus `/tmp/nvim_python.log`)?  Also, I would have expected `:py print 1+1` to error for Python 3, so I don't think you tested the right version.  You may want to modify it to `:py3 print(1+1)` and see if anything unusual shows up.

**Comment by PonderingGrower** [comment]
Ok, so I've set `let g:python3_host_prog = '/usr/bin/python3.3'` but I get the same error.

But you were right. When I do `:py3 print(1+1)` I get:

```
Vim(if):Channel was closed by the client                                                                                                                                                                                                       
Failed to load Python host. You can try to see what happened by starting Neovim with the environment variable $NVIM_PYTHON_LOG_FILE set to a file and opening the generated log file. Also, the host stderr will be available in Neovim log, so
 it may contain useful information. See also ~/.nvimlog. 
```

Nothing in `~/.nvimlog`.

**Comment by jszakmeister** [comment]
Okay.  Unfortunately, you've hit the limit of my knowledge on the python integration.  Hopefully someone else can jump in and help you.  Sorry!

**Comment by PonderingGrower** [comment]
Thanks a bunch for helping me dig deeper!

**Comment by PonderingGrower** [comment]
I have added the following line above [this line](https://github.com/neovim/neovim/blob/master/runtime/autoload/remote/host.vim#L211):
```
echomsg 'Channel ID: '. channel_id
```
And the output I get when I try `:py3 print(1+1)` is: `Channel ID: 1`.

**Comment by mhinz** [comment]
@fwalch @bfredl Any ideas?

**Comment by vito-c** [comment]
@PonderingGrower I would try it without any plugins installed first. Then try it with unite.vim. I was having issues with unite.vim's remote api at one point. You could try updating your plugins again as well.
`nvim -u vimrc-minimal`
then `:UpdateRemotePlugins`

vimrc.minimal
```
set expandtab
syntax on
set mouse=
let mapleader = ' '
```

**Comment by PonderingGrower** [comment]
I tried the vimrc you suggested and I get the same thing when I run `:py3 print(1+1)`:

```
Vim(if):Channel was closed by the client
Failed to load Python host. You can try to see what happened by starting Neovim with the environment variable $NVIM_PYTHON_LOG_FILE set to a file and opening the generated log file. Also, the host stderr will be available in Neovim log, so
 it may contain useful information. See also ~/.nvimlog.
```

When i run `:UpdateRemotePlugins` nothing(visible) happens.

**Comment by critiqjo** [comment]
>That looks like Python support is missing.

I have 2 remote plugins: one which optionally requires python3 client (Unite), and another which definitely requires python2 client (lldb.nvim). If I only have python2 client, I *should* be able to get lldb.nvim working, but `:UpdateRemotePlugins` errors midway without registering *any* remote plugins. I believe this is not the desired behavior:
```
Vim(let):E903: Could not spawn API job
Error detected while processing function <SNR>76_UpdateRemotePlugins..<SNR>76_
RegistrationCommands..remote#host#Require..<SNR>76_RequirePythonHost:
line   21:
E605: Exception not caught: Failed to load Python host. You can [...] See also ~/.nvimlog.

Error detected while processing function <SNR>76_UpdateRemotePlugins..<SNR>76_
RegistrationCommands..remote#host#Require:
line   10:
E171: Missing :endif

Error detected while processing function <SNR>76_UpdateRemotePlugins:
line    5:
E171: Missing :endif
```

**Comment by vito-c** [comment]
@critiqjo try with `unite` disabled and try with no plugins.

**Comment by critiqjo** [comment]
@vito-c Of course it'll work if I disable all plugins. That's the point: `:UpdateRemotePlugins` should be more robust, and work gracefully regardless of whether there are misbehaving plugins/remote clients.

**Comment by vito-c** [comment]
@critiqjo I was trying to help you find out which plugin is causing the issue I suspect you might not have python3 support fully setup

**Comment by critiqjo** [comment]
@vito-c That is correct. I intentionally did not install python3 client, and Unite is the plugin causing the issue.

**Comment by wsdjeg** [comment]
i have got the same issue ,that is why i can not use py3 plugin
@DonnieWest

**Comment by Shougo** [comment]
Oh, it is the fatal problem.  I will fix it.
Please wait.

**Comment by wsdjeg** [comment]
@PonderingGrower @vito-c @critiqjo 
i have fix this issue,

**Comment by wsdjeg** [comment]
if you install python3.4 and python 3.5 together ,just make sure same version being used by nvim and shell

**Comment by Shougo** [comment]
Yes, but I will fix @critiqjo reported problem.

**Comment by wsdjeg** [comment]
hi @Shougo i say i have fix the issue with UpdateRemotePlugins,but gundo's issue still exists

**Comment by wsdjeg** [comment]
before change my python version i got the same issue when use UpdateRemotePlugins

**Comment by wsdjeg** [comment]
now i can use it 
![2015-12-07 22-03-35](https://cloud.githubusercontent.com/assets/13142418/11628711/6797a40e-9d2e-11e5-9298-b7bde9f7322c.png)

**Comment by Shougo** [comment]
> hi @Shougo i say i have fix the issue with UpdateRemotePlugins,but gundo's issue still exists

It is neovim-python problem.
Please create new issue in it.  I will check it.
But please explain to reproduce ways.  Your report information is not enough to reproduce it.

**Comment by wsdjeg** [comment]
ok i create an new issue reproduce all the thing

**Comment by Shougo** [comment]
@critiqjo I fixed your reported problem.  Please check it.

**Comment by justinmk** [comment]
Assuming this is addressed by  https://github.com/neovim/neovim/pull/3813

**Comment by revolter** [comment]
Can somebody help me with https://github.com/critiqjo/lldb.nvim/issues/46?

**Comment by vito-c** [comment]
@revolter what version of neovim? 

Looks like you are using mac os sierra? Have you installed the command line tools in xcode?
`xcode-select --install`

I just installed lldb plugin to try it out. Make sure you follow these instructions. http://blog.rplasil.name/2016/03/how-to-debug-neovim-python-remote-plugin.html

Specifically make you do
`sudo /System/Library/Frameworks/Python.framework/Versions/2.7/bin/python -m easy_install neovim`
and add this to your vimrc:
`let g:python_host_prog = '/System/Library/Frameworks/Python.framework/Versions/2.7/bin/python'`

**Comment by revolter** [comment]
@vito-c, Version `0.1.7`. I did have command line tools installed.

I thought `pip install neovim` would suffice :( I can't understand how I missed that step 3. Do you want the money or should I donate them to Neovim?
